components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,278 @@
+            playing_effect_promise.reject_error(Error::NotSupported);
+        }
+
+        *self.playing_effect_promise.borrow_mut() = Some(playing_effect_promise.clone());
+        let play_effect_timestamp = self.global().performance().Now();
+
+        let params = DualRumbleEffectParams {
+            duration: params.duration as f64,
+            start_delay: params.startDelay as f64,
+            strong_magnitude: *params.strongMagnitude,
+            weak_magnitude: *params.weakMagnitude,
+        };
+        let event = EmbedderMsg::PlayGamepadHapticEffect(
+            self.gamepad_index as usize,
+            embedder_traits::GamepadHapticEffectType::DualRumble(params),
+        );
+        self.global().as_window().send_to_embedder(event);
+
+        playing_effect_promise
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dom-gamepadhapticactuator-reset>
+    fn Reset(&self) -> Rc<Promise> {
+        let reset_result_promise = Promise::new(&self.global());
+
+        let document = self.global().as_window().Document();
+        if !document.is_fully_active() {
+            reset_result_promise.reject_error(Error::InvalidState);
+        }
+
+        if self.playing_effect_promise.borrow().is_some() {
+            *self.reset_result_promise.borrow_mut() = Some(reset_result_promise.clone());
+
+            let event = EmbedderMsg::StopGamepadHapticEffect(self.gamepad_index as usize);
+            self.global().as_window().send_to_embedder(event);
+        } else {
+            let message = DOMString::from("complete");
+            reset_result_promise.resolve_native(&message);
+        }
+
+        reset_result_promise
+    }
+}
+
+impl GamepadHapticActuator {
+    pub fn has_playing_effect_promise(&self) -> bool {
+        self.playing_effect_promise.borrow().is_some()
+    }
+
+    pub fn resolve_playing_effect_promise(&self) {
+        let playing_effect_promise = self.playing_effect_promise.borrow().clone();
+        if let Some(promise) = playing_effect_promise {
+            let message = DOMString::from("complete");
+            promise.resolve_native(&message);
+        }
+    }
+
+    pub fn has_reset_result_promise(&self) -> bool {
+        self.reset_result_promise.borrow().is_some()
+    }
+
+    pub fn resolve_reset_result_promise(&self) {
+        let playing_effect_promise = self.playing_effect_promise.borrow().clone();
+        let reset_result_promise = self.reset_result_promise.borrow().clone();
+
+        if playing_effect_promise.is_some() {
+            let trusted_promise = TrustedPromise::new(
+                self.playing_effect_promise
+                    .borrow()
+                    .clone()
+                    .expect("Promise is null!"),
+            );
+            let _ = self.global().gamepad_task_source().queue(
+                task!(preempt_promise: move || {
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("preempted");
```

Comment:
This should only fire if the promise hasn't already been preempted by a playEffect call after the reset call that ended-up queuing this task(see comment above).